File: scripts/fix-security-regex.js
#!/usr/bin/env node

/**
 * Fix Regex Issues in security.ts
 * Fixes three double-backslash regex patterns
 */

const fs = require('fs');
const path = require('path');

// Colors for console output
const colors = {
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m'
};

function log(color, symbol, message) {
  console.log(`${colors[color]}${symbol}${colors.reset} ${message}`);
}

console.log('');
console.log(`${colors.blue}════════════════════════════════════════${colors.reset}`);
console.log(`${colors.blue}   Fix security.ts Regex Issues${colors.reset}`);
console.log(`${colors.blue}════════════════════════════════════════${colors.reset}`);
console.log('');

const filePath = path.join(__dirname, '..', 'supabase', 'functions', 'server', 'security.ts');

// Check if file exists
if (!fs.existsSync(filePath)) {
  log('red', '✗', `Error: ${filePath} not found!`);
  process.exit(1);
}

// Read the file
let content = fs.readFileSync(filePath, 'utf8');

// Create backup
const backupPath = `${filePath}.backup.${Date.now()}`;
fs.writeFileSync(backupPath, content, 'utf8');
log('green', '✓', `Backup created: ${path.basename(backupPath)}`);
console.log('');

// Count issues before
const issuesBefore = (content.match(/\\\\w|\\\\D|\\\\\./g) || []).length;
log('yellow', '→', `Found ${issuesBefore} regex issues with double backslashes`);
console.log('');

// Fix 1: String sanitization - on\\w+= should be on\w+=
log('yellow', '→', 'Fixing: .replace(/on\\\\w+=/gi, \'\') → .replace(/on\\w+=/gi, \'\')');
content = content.replace(/\/on\\\\w\+=/g, '/on\\w+=/');

// Fix 2: Email validation - \\. should be \.
log('yellow', '→', 'Fixing: Email regex \\\\. → \\.');
content = content.replace(/\\\\\\\./g, '\\.');

// Fix 3: Phone validation - \\D should be \D
log('yellow', '→', 'Fixing: .replace(/\\\\D/g, \'\') → .replace(/\\D/g, \'\')');
content = content.replace(/\/\\\\\\\\D\//g, '/\\D/');

console.log('');

// Write the fixed content
fs.writeFileSync(filePath, content, 'utf8');

// Count issues after
const issuesAfter = (content.match(/\\\\w|\\\\D|\\\\\./g) || []).length;

if (issuesAfter === 0) {
  log('green', '✓', 'All regex issues fixed!');
  console.log('');
  console.log('Changes made:');
  console.log('  - Fixed on\\w+= pattern in string sanitization');
  console.log('  - Fixed \\. pattern in email validation');
  console.log('  - Fixed \\D pattern in phone validation');
  console.log('');
  console.log(`${colors.green}════════════════════════════════════════${colors.reset}`);
  console.log(`${colors.green}   Success! Ready to redeploy${colors.reset}`);
  console.log(`${colors.green}════════════════════════════════════════${colors.reset}`);
  console.log('');
  console.log('Next steps:');
  console.log('  1. Deploy to dev: ./scripts/redeploy-backend.sh dev');
  console.log('  2. Test login at: http://localhost:5173/admin/login');
  console.log('');
  console.log('To restore backup if needed:');
  console.log(`  cp ${backupPath} ${filePath}`);
  console.log('');
} else {
  log('red', '✗', `Warning: ${issuesAfter} issues still remain`);
  console.log('');
  
  // Show remaining issues
  const lines = content.split('\n');
  lines.forEach((line, index) => {
    if (/\\\\w|\\\\D|\\\\./.test(line)) {
      console.log(`  Line ${index + 1}: ${line.trim()}`);
    }
  });
  
  console.log('');
  console.log('You can restore the backup if needed:');
  console.log(`  cp ${backupPath} ${filePath}`);
  process.exit(1);
}

File: scripts/fix-security-regex.py
#!/usr/bin/env python3

"""
Fix Regex Issues in security.ts
Fixes three double-backslash regex patterns
"""

import os
import re
import sys
from datetime import datetime
from pathlib import Path

# Colors for console output
class Colors:
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    RESET = '\033[0m'

def log(color, symbol, message):
    print(f"{color}{symbol}{Colors.RESET} {message}")

def main():
    print()
    print(f"{Colors.BLUE}═══════════════════════════════════════={Colors.RESET}")
    print(f"{Colors.BLUE}   Fix security.ts Regex Issues{Colors.RESET}")
    print(f"{Colors.BLUE}═══════════════════════════════════════={Colors.RESET}")
    print()
    
    # Get file path
    script_dir = Path(__file__).parent
    file_path = script_dir.parent / 'supabase' / 'functions' / 'server' / 'security.ts'
    
    # Check if file exists
    if not file_path.exists():
        log(Colors.RED, '✗', f'Error: {file_path} not found!')
        sys.exit(1)
    
    # Read the file
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Create backup
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    backup_path = file_path.parent / f'security.ts.backup.{timestamp}'
    with open(backup_path, 'w', encoding='utf-8') as f:
        f.write(content)
    log(Colors.GREEN, '✓', f'Backup created: {backup_path.name}')
    print()
    
    # Count issues before
    issues_before = len(re.findall(r'\\\\w|\\\\D|\\\\\\.', content))
    log(Colors.YELLOW, '→', f'Found {issues_before} regex issues with double backslashes')
    print()
    
    # Fix 1: String sanitization - on\\w+= should be on\w+=
    log(Colors.YELLOW, '→', 'Fixing: .replace(/on\\\\w+=/gi, \'\') → .replace(/on\\w+=/gi, \'\')')
    content = content.replace('/on\\\\w+=/gi', '/on\\w+=/gi')
    
    # Fix 2: Email validation - \\\. should be \.
    log(Colors.YELLOW, '→', 'Fixing: Email regex \\\\. → \\.')
    content = content.replace('\\\\.', '\\.')
    
    # Fix 3: Phone validation - \\D should be \D
    log(Colors.YELLOW, '→', 'Fixing: .replace(/\\\\D/g, \'\') → .replace(/\\D/g, \'\')')
    content = content.replace('/\\\\\\\\D/', '/\\D/')
    
    print()
    
    # Write the fixed content
    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)
    
    # Count issues after
    issues_after = len(re.findall(r'\\\\w|\\\\D|\\\\\\.', content))
    
    if issues_after == 0:
        log(Colors.GREEN, '✓', 'All regex issues fixed!')
        print()
        print('Changes made:')
        print('  - Fixed on\\w+= pattern in string sanitization')
        print('  - Fixed \\. pattern in email validation')
        print('  - Fixed \\D pattern in phone validation')
        print()
        print(f"{Colors.GREEN}═══════════════════════════════════════={Colors.RESET}")
        print(f"{Colors.GREEN}   Success! Ready to redeploy{Colors.RESET}")
        print(f"{Colors.GREEN}═══════════════════════════════════════={Colors.RESET}")
        print()
        print('Next steps:')
        print('  1. Deploy to dev: ./scripts/redeploy-backend.sh dev')
        print('  2. Test login at: http://localhost:5173/admin/login')
        print()
        print('To restore backup if needed:')
        print(f'  cp {backup_path} {file_path}')
        print()
    else:
        log(Colors.RED, '✗', f'Warning: {issues_after} issues still remain')
        print()
        
        # Show remaining issues
        lines = content.split('\n')
        for i, line in enumerate(lines, 1):
            if re.search(r'\\\\w|\\\\D|\\\\\\.', line):
                print(f'  Line {i}: {line.strip()}')
        
        print()
        print('You can restore the backup if needed:')
        print(f'  cp {backup_path} {file_path}')
        sys.exit(1)

if __name__ == '__main__':
    main()
